### ISSUE TITLE
Highway environment for DQN agents

### ISSUE THREAD

**Issue by waterhorse1** [eleurent/highway-env#38]
Thank you for your efforts for this great environment !

I'm training a DQN agents on highway-v0 environments. I have conducted some experiments based on your setting in rl_agents/highway/DQN_agents for about 1k episodes but still failed to get stable DQN agents. So I wonder how many episodes the training for DQN agents will take under the settings in that directory to achieve great performance similar with what you show in the video.  It will help if you can detail your training settings.

In addition, the simulation for each episode is kind of slow for training. So Does there exist some tools to accelerate the simulation if I don't render the environments for faster training.(usually run DQN for 1k episodes will take 1h)

**Comment by eleurent** [comment]
Hi @waterhorse1, thanks!
I think I used 1k episodes for training (and I agree that this parameter should appear in the agent configuration). Regarding performances, I think the current `baseline.json` configuration is the one I used. I will try to reproduce it soon and see if I notice anything wrong.

Regarding speed, I agree that `highway-env` is not particularly well optimised and that simulating 50 vehicles only can be quite slow (see https://github.com/eleurent/highway-env/issues/36 , the default highway-v0 configuration is 2x real-time only). The easiest way to adress that is to reduce the number of vehicles on the road. And with fewer vehicles to pass, the episode duration can also be reduced.
You can use the environment configuration `rl-agents/scripts/configs/HighwayEnv/env_medium.json` that uses 20 vehicles and 30s duration, which increases the real-time ratio to x10), and should not affect the test performance too much.

**Comment by waterhorse1** [comment]
Thank you for your reply! 

In fact, the average reward is increasing with training process going on so I think your DQN code is correct. There just exists some extremely bad cases ended with vehicle collisions. So I also doubt whether that is due to the lack of training or DQN agents just cannot show stable performance like what planning algorithms can do. Hope you can reproduce that soon and give more feedback.

**Comment by waterhorse1** [comment]
Sorry to disturb you. I have noticed that you add some new changes to dqn code(replace default dqn loss setting from smooth-l1 to l2). I have tried that on highway environment but still failed to get stable rl agents. Can you help me with reproduce dqn agent on highway environment? Thanks a lot!

**Comment by eleurent** [comment]
Hi @waterhorse1, I updated the default loss function so that CartPole works with a minimalist configuration, but I haven't check which loss works best for highway-env. If smooth L1 is better than L2, it should appear in the highway configuration.

I'm sorry for the delay, I've been and am still going to be quite busy this week,  but I'll try to run these experiments next week and get back to you then.

What do you expect exactly from "stable" performance ? Reaching an optimal performance ? Having a non-decreasing average reward during training, that converges to a constant final performance ? I think that the latter is not guaranteed by Deep RL algorithms, even with DQN on a simple cartpole you can get a sudden collapse of performance after having reached optimality. People often use "early stopping" to hide such effects (i.e. only show the graph of rewards before the collapse), but I think it is actually quite common.

**Comment by waterhorse1** [comment]
Thank you for your reply. In fact, I just failed to get non-decreasing average reward during training, which converges to a constant final performance. Will appreciate it if you can run some tests regarding that.

**Comment by waterhorse1** [comment]
Wow, I have redownloaded your code and retest highway environment based on that. The DQN agents seem great after that. And I also tested l1/l2 loss on that environment and found l2 loss show much better performance. So maybe the configuration for that should be changed to l2 loss.

**Comment by eleurent** [comment]
Thanks for this feedback, that's interesting!
The loss is not specified in the [highway baseline](https://github.com/eleurent/rl-agents/blob/master/scripts/configs/HighwayEnv/agents/DQNAgent/baseline.json) configuration, so the default loss is used.
I guess that at some point, for some reason, I changed the default from L2 to L1 while working on some other project, and did not notice the drop in performance. As you noticed I recently switched the default back to L2 which seems generally better. 
I agree that these configurations should not rely too much on default values. Now that you have confirmed that L2 > L1 for this particular env at least, this should be in the config.

**Comment by eleurent** [comment]
Also, I suggest you try the `HighwayEnv/agents/DQNAgent/ego_attention.json` agent configuration along with the `HighwayEnv/env_obs_attention.json` environment configuration, which I found learns and performs better than the baseline vanilla MLP.

**Comment by Derekabc** [comment]
Hi author, Sorry for bothering you! I tested the Highwayenv using A2C, the algorithm converges to reward 30 after thousands of training. While when I tested the learned agent, I found the problem is the same as described in https://github.com/eleurent/rl-agents/issues/21, the video shows the agent does not overtake, speed up, or change lanes. I am not sure of is that I choose the wrong parameters while it did reach a high reward.

Also, for reward 30, how is it calculated. duration is 40. while high_speed_reward = 0.4 and keep_right = 0.1. I cannot understand why we can get about 30 for 40-time steps.

Thanks for your time and insights!

**Comment by eleurent** [comment]
Hi @Derekabc,,
Indeed, it seems that something went wrong somewhere if you do not manage to get a good policy with A2C. Unfortunately, there could be many reasons for that and I don't really know how to help you. I advise you to check whether highway-env behave as intended (actions yield the proper effect, rewards are higher when driving faster, 0 at collisions, etc), and then investigate your learning algorithm.

Regarding the reward 30 for 40 steps, the coefficients you mentioned are correct (with a penalty of -1 for collisions), but the reward are also normalized in [0, 1] by a linear mapping from the original [-1, 0.4+0.1] range.
